### DISCUSSION TITLE
MIT 6.S081: Operating System Engineering - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#106] [Announcements]
# MIT 6.S081: Operating System Engineering - CS自学指南

CS自学指南

https://csdiy.wiki/%E6%93%8D%E4%BD%9C%E7%B3%BB%E7%BB%9F/MIT6.S081/

**Comment by PorunC** [comment]
个人整理的一些S081能用到的博客和资料，希望能为这份自学指南做点贡献，另外有关  xv6-riscv 操作系统的学习可以看深圳大学这位老师发在 哔哩哔哩 的视频：[正明大佐](https://space.bilibili.com/1040264970)
## 参考资料
- [doraemonzzz](http://doraemonzzz.com/tags/6-S081/)
- [Xiao Fan (樊潇)](https://fanxiao.tech/posts/MIT-6S081-notes/)
- [Miigon's blog](https://blog.miigon.net/tags/operating-system/)
- [Zhou Fang](https://walkerzf.github.io/categories/6-S081/index.html)
- [Yichun's Blog](https://www.yichuny.page/tags/Operating%20System)
- [解析Ta](https://blog.csdn.net/u013577996/article/details/108679997)
- [PKUFlyingPig](https://github.com/PKUFlyingPig/MIT6.S081-2020fall)
- [星遥见](https://www.cnblogs.com/weijunji/tag/XV6/)

**Comment by Jacob-liu1996** [comment]
自荐一波，b站持续更新，xv6源码与lab讲解：https://www.bilibili.com/video/BV1ou41127p9/
补充一下，教材里面的另一位作者Russ Cox是golang语言的创作者之一，也是巨佬级别的人物

**Reply by 1000charsdance** [reply]
2022/7/8日开始做，四处碰壁，快放弃了，昨晚睡觉前在b站找到楼主的视频添加到稍后再看，今天想起这个wiki，进来一看，咦？这不是我昨晚找到的视频吗？跟着视频做，现在已经搭建好了环境，用的是wls2的ubuntu，感谢！
我认为换apt源和./configure这两点比较重要，想学的看这位大哥的视频应该没问题！

**Reply by tuyula** [reply]
一起来学习么, 微信tuyu550574002

**Comment by Junxin-L** [comment]
推荐一下OSTEP，是一个在线的电子教材，写的也很清楚，很好入门~

**Reply by kyangcmXF** [reply]
OSTEP就是NJU OS 2021,2022的课程教材。应该说OSTEP是top级别的操作系统教科书了

**Comment by aswrise** [comment]
推荐一份翻译笔记：
https://mit-public-courses-cn-translatio.gitbook.io/mit6-s081/

**Reply by jiuyueshenhua** [reply]
进不去了

**Reply by roboatLee** [reply]
新字幕
https://github.com/mayf09/6.S081-2020-video-subtitles

**Reply by clownorprime** [reply]
https://github.com/MagicMarvel/MIT6.S081-translate

**Reply by roboatLee** [reply]
https://github.com/huihongxiao/MIT6.S081
gitbook有时打不开，可以下载这个，是一模一样的

**Comment by fung-hwang** [comment]
推荐一份非常棒的中文资料 [http://xv6.dgs.zone/](http://xv6.dgs.zone/)

**Comment by linepi** [comment]
有人知道如何注册piazza吗？

**Reply by linepi** [reply]
欢迎交流wechat id: 18481265082

**Reply by linepi** [reply]
自荐魔改xv6仓库https://github.com/linepi/xv6，实现了很多功能（环境变量，动态栈，backtrace打印文件行）。欢迎交流～

**Reply by Vossera** [reply]
https://piazza.com/mit/fall2021/6s081

**Comment by Vanilla-Beauty** [comment]
想问下，学这个需要先学riscv汇编吗？我只通过csapp学过x86汇编

**Reply by linepi** [reply]
riscv非常简单，学过x86再学就是换了个皮

**Comment by yifan0414** [comment]
网站打不开了吗

**Comment by xinghyfish** [comment]
课程网站打不开了QAQ

**Comment by linyn-zero** [comment]
推荐一份S081资源共享文档https://tarplkpqsm.feishu.cn/docs/doccnBFsXFMsAr1oXEVsaT9E3Jg#

**Comment by taylover2016** [comment]
昨天正式开始看，今天 Lab 写了一天。刚开始 Easy(Morris 大佬给出的预计时间是小于 1 小时) 的题目都写了大半天... 好在最后调出来了。确实自己写会理解深入不少，也会发现很多问题。

**Comment by Ywinh** [comment]
请问这门课的video值得听吗，草率浏览了一下全是板书，我的想法是只看翻译笔记(https://mit-public-courses-cn-translatio.gitbook.io/mit6-s081/)

**Reply by Wolves754** [reply]
如果你需要做实验，我感觉尽量还是跟着视频好一些

**Reply by krmmzs** [reply]
我是感觉自己不适合看视频, 直接看的翻译笔记(野生note), 感觉没啥问题.
其中对一些学生问题进行位置重排感觉挺好的.

**Reply by sywx-jx** [reply]
打不开了，请问还有什么办法获取

**Comment by hzhang-coding** [comment]
这个和cs162推荐哪个呢

**Reply by AlexMan2000** [reply]
有很多时间的话就cs162(一个人独立完成难度太高太高), 时间紧就mit6.s081

**Comment by murongke83** [comment]
LEC2 youtube 视频上没有，有没有同学补一下！？

**Comment by sarendis56** [comment]
页面里超链接的教程已经更新到rev3了: https://pdos.csail.mit.edu/6.828/2023/xv6/book-riscv-rev3.pdf

**Reply by n-WN** [reply]
https://pdos.csail.mit.edu/6.828/2024/xv6/book-riscv-rev4.pdf

**Comment by abinzzz** [comment]
组队学习，相互交流+q：1589036128

**Comment by cubaneOvO** [comment]
MIT6.1810（6.S081的2022版本) Lab全解 链接：https://juejin.cn/column/7276350321094082614
欢迎大家来交流 ！！！

**Comment by Aventadoor8003** [comment]
能问一下测试怎么办吗？是不是只能自己想办法自己测了？

**Reply by linepi** [reply]
有测试的

**Reply by King2Jack** [reply]
可以请教下怎么测试吗？我这好像没办法make grade

**Reply by yali-hzy** [reply]
可以看一下自己有没有交叉编译环境，如果按照实验说明用apt安装不行的话，可以去github上找官方仓库自己编译一下
https://github.com/riscv-collab/riscv-gnu-toolchain

**Comment by kyzyc** [comment]
个人看法，Robert 教授讲的更清楚一点，很多时候我的疑问刚刚产生他就预知给你解决了。另一位Kaashoek教授能感觉到他是很懂的，但是感觉表述起来容易让人云里雾里，有点站在上帝视角发言的感觉，再加上口音........

**Comment by LysanderT** [comment]
想知道大家只用20和21年两个版本的原因是？

**Reply by linepi** [reply]
没啥原因，可能是答案比较多hh

**Reply by ShelpAm** [reply]
只找得到2020年的课程视频了，我也想看最新的

**Comment by Coodist** [comment]
强烈推荐b站up Deniffer的环境配置教程，感觉这位大佬的教程是最详尽的
https://www.bilibili.com/video/BV11K4y127Qk
这是他在知乎上的个人文章https://zhuanlan.zhihu.com/p/341264783
如果还是解决不了可以加我q211699660，因为我知道配环境到处Error的过程真的很折磨人

**Comment by koarz** [comment]
有人做了2023的吗，我的Lab3vmprint()不打印509那个pagetable，我在网上找到各种答案也都打印不出509，所以应该不是我的问题吧

**Reply by kj5377707** [reply]
你要先做第一個再做print 才會有509這個page

**Reply by koarz** [reply]
哦哦，谢谢，知道了

**Comment by mrmign** [comment]
6.S081与 [6.1810](https://pdos.csail.mit.edu/6.1810/2023/schedule.html)啥区别？

**Reply by yijun-tang** [reply]
是同一门课，换了课程编号

**Comment by ClimberLeo** [comment]
我想问下如果要做xv6lab的话是要跟着他的课学才好吗？可以看OSTEP这种书学然后做吗？

**Comment by alehpwind** [comment]
准备跟着2023版本学，刚开始，有没有群可以加啊，想找人一起。我qq：2488132688

**Reply by King2Jack** [reply]
有2023版本的网址吗？

**Comment by SMHILYZYH** [comment]
【The xv6 Kernel-哔哩哔哩】 https://b23.tv/gk4KaoK
可以看看这个，直接讲解xv6的代码，上传了中英文字幕，感觉还不错

**Comment by ShaneHeX** [comment]
[6.S081](https://github.com/hexinatgithub/xv6-labs-2020/tree/d3542e509bc2120b3a3128f851ffaacb83d16982) lab 全部完成，测试通过

**Comment by asdfefbhjy** [comment]
https://github.com/OccupyMars2025/xv6-labs-2023
2024/2/1开始学习，目标: 2024/4/30之前彻底学完，欢迎提PR或者issue交流

**Comment by LCJD99** [comment]
## 找同学一起学习
刚学习完**cs61a**, **csapp**的课程，计划后续还会学习其他课程(目前是**cs61b** 和 **mit6.s081**)。

创建了一个qq学习群，有一起学习的同学可以加群一起讨论，qq群： **563729367**

**Comment by mzshuo** [comment]
mmap那个lab有人做了共享内存的challenge吗？

**Comment by dghtucs** [comment]
建了一个QQ课程群，想讨论可入758415268

**Comment by Nicooo-Wang** [comment]
汇编有什么好的公开课吗，最好是risc v

**Comment by Nicooo-Wang** [comment]
riscv汇编怎么系统学习呢？有教程吗？

**Reply by ClimberLeo** [reply]
我记得ddca有教的，找riscv版本的就行，而且如果只是做6s081的话你一边做一边看这他给的riscv官方文档学也可以的，我就是这样，不懂的再问gpt，主要riscv太简洁了，也很好学，文档才一二百页

**Reply by Nicooo-Wang** [reply]
太狠了老哥，从软件直接干到硬件设计。。。
谢谢，我之后有空了学习一下，确实CPU 流水线这些东西也很重要

**Reply by ClimberLeo** [reply]
哦对，你不要误会我的意思了。。。我的建议还是如果你只是想学6s081的话，一边看文档查GPT一边做完全可以的，riscv无非就是一套指令集，常用的指令也就那些，并且很精简，毕竟硬件电路那些东西学起来也有成本，完全陷进去就有点舍本逐末的意思了，但你如果对硬件电路体系架构之类感兴趣并且有时间的话可以去学学，除了ddca外还有一本计算机组成与设计riscv版也可以，微信读书上有

**Reply by xiaofengwang01** [reply]
可以看看berkeley的cs61c课程，做完与riscv相关的两个lab和一个project，你就会对riscv非常了解了

**Comment by King2Jack** [comment]
请教一下，从github下载的仓库里，没有带自动评测工具吗？

**Reply by ShelpAm** [reply]
有的

**Comment by ytoso** [comment]
2021版本的实验少了lazy allocation，大家觉得做了COW还有必要做lazy吗

**Reply by Tinuvile** [reply]
老哥现在有无看法

**Comment by q2333gh** [comment]
补充资料:  6s081 all-in-one https://xv6.dgs.zone/

**Comment by AlexMan2000** [comment]
有没有正在学的同学在debug xv6的时候遇到stepi 进不了kernel space的情况啊(stepi ecall 直接跳到ret了)， 本人使用WSL Ubuntus 22.04的环境

**Reply by Lupinus** [reply]
同，请问解决了吗

**Reply by Vanilla-Beauty** [reply]
用 20.04, qemu用指定的版本

**Comment by fjc-hub** [comment]
**20天轻松拿下**， 最难最tricky的实验应该是Lab lock

**Reply by moyuhualuo** [reply]
佬，学的那个版本，最近想要自学下

**Reply by fjc-hub** [reply]
> 佬，学的那个版本，最近想要自学下

2022版本，因为我的WSL Ubuntus 22.04机器配置2021版本的实验环境有点麻烦

**Comment by kingcanfish** [comment]
2024 版的syscall lab 实验 新出了一个attack lab 有朋友有思路吗～

**Reply by Lupinus** [reply]
老哥你系统调用那里的调试和教授课上的一样吗，为什么我有好多不一致的地方

**Reply by kingcanfish** [reply]
> 老哥你系统调用那里的调试和教授课上的一样吗，为什么我有好多不一致的地方

@Lupinus 我没怎么看课程 直接做实验的

**Reply by yali-hzy** [reply]
找字符串的时候，逐字符匹配，只要相同部分超过串长一半就认为是找到了。或者你调试的时候，把分配的虚页和物理页地址输出一下，比对一下地址也可以发现每次运行时attack对应的虚页地址都是相同的

**Reply by kingcanfish** [reply]
> 找字符串的时候，逐字符匹配，只要相同部分超过串长一半就认为是找到了。或者你调试的时候，把分配的虚页和物理页地址输出一下，比对一下地址也可以发现每次运行时attack对应的虚页地址都是相同的

哈哈哈是的，后面做出来了，原因是字符串的前一半的那部分归还给伙伴系统的时候用作链表节点了 所以覆盖了 后面一半是能比较出来的

**Comment by AlexMan2000** [comment]
同问，我用window subsystem linux 22.04 调试的时候进不了kernel space


Xinyu Kuo ***@***.***> 于2024年11月22日周五 02:33写道：

**Reply by Vanilla-Beauty** [reply]
用20.04

**Comment by LRL52** [comment]
2024.12.8 我完成了 xv6 labs 2023 版。自荐一下我的 Lab1-10 的[个人笔记](https://lrl52.top/category/6-1810/)（从配环境启动、教材笔记摘录到到完成最后一个实验），以及[个人代码仓库](https://github.com/LRL52/xv6-labs-2023)（内含学习和完成实验用到的所有文档资料），原创代码实现，码风良好，有英文注释，希望能对后人有所帮助~😘

**Comment by Anekoique** [comment]
2024的Lab3 page table中print pgtbl给出的答案是错的

**Reply by shinnenijou** [reply]
如果你说的是这个地方给的示例的话，
\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=
When you run print_kpgtbl() test, your implementation should print the following output:
/\* ... \*/
..0xffffffffc0000000: pte 0x0000000021fc8401 pa 0x0000000087f21000
/\* ... \*/
\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=\=
最后几个0xffffffffc0000000感觉确实是有问题，这个地址已经超过了 kernel/riscv.h 里定义的MAXVA，虽然RISC-V也允许内核使用虚拟地址没有用于地址转换的位作为内核区域和用户区域的区分，但是xv6没有这么使用而是直接都放在了0～MAXVA范围内

**Reply by Anekoique** [reply]
> 如果你说的是这个地方给的示例的话， ========================= When you run print_kpgtbl() test, your implementation should print the following output: /* ... */ ..0xffffffffc0000000: pte 0x0000000021fc8401 pa 0x0000000087f21000 /* ... */ ========================= 最后几个0xffffffffc0000000感觉确实是有问题，这个地址已经超过了 kernel/riscv.h 里定义的MAXVA，虽然RISC-V也允许内核使用虚拟地址没有用于地址转换的位作为内核区域和用户区域的区分，但是xv6没有这么使用而是直接都放在了0～MAXVA范围内

是的，我还复现了为什么没有打印出正确的值，他在对参数做移位的时候发生了溢出

**Reply by shinnenijou** [reply]
> 是的，我还复现了为什么没有打印出正确的值，他在对参数做移位的时候发生了溢出

拜见了一下你的代码，示例是不是这个原因我不清楚，你的那个其实不完全是因为i位移溢出，即使没有溢出，左移以后最高位是1了的话，int向uint64扩展的过程里先符号拓展到long，拓展出去的位数就全是0xffff...了（最后i声明为uint64的修改是对的，如果声明成uint32就有溢出的问题，最后会变成全是0x00000000c0000000这样的数字）

**Reply by Anekoique** [reply]
> > 是的，我还复现了为什么没有打印出正确的值，他在对参数做移位的时候发生了溢出
> 
> 拜见了一下你的代码，示例是不是这个原因我不清楚，你的那个其实不完全是因为i位移溢出，即使没有溢出，左移以后最高位是1了的话，int向uint64扩展的过程里先符号拓展到long，拓展出去的位数就全是0xffff...了（最后i声明为uint64的修改是对的，如果声明成uint32就有溢出的问题，最后会变成全是0x00000000c0000000这样的数字）

是的，非常严谨👍

**Comment by masterAllen** [comment]
2024年的locks第一个问题，memory allocator，我发现我的 test3 经常会失败，大家有遇到这种情况吗？

**Comment by bowenyu066** [comment]
现在这门课改名叫6.1810了：https://pdos.csail.mit.edu/6.828/2024/index.html

**Comment by copicomi** [comment]
给后来的提个醒，直接做lab可能不好下手，有的实验必须读内核源码才能理解机制，建议是跟课程官网的阅读材料先过一遍内容和源码

**Comment by zhipzhang** [comment]
微信 astro_zhipzhang， 看看有没有小伙伴一起

**Reply by siolyn** [reply]
我还没到一半，可以讨论一下吗

**Comment by Nicoledon** [comment]


**Reply by ShelpAm** [reply]
有的

**Comment by Zlatanwic** [comment]
请问有24年的视频吗

**Comment by vixitor** [comment]
有没有人一起做，目前做到lab2，attack遇到点问题

**Reply by vixitor** [reply]
我的微信William_hth

**Comment by JianOUO** [comment]
推荐一个xv6 book的繁体中文翻译：https://mes0903.github.io/OS/xv6-riscv-book-zh-TW/
翻译流畅，排版舒适，对于困难的内容作者还提供了贴心的注释

**Comment by SweetGargamel** [comment]
想问一下这个和jyyos的区别是啥啊，如果已经学过CSAPP,且做过南大PA实验【虽然不全是自己写的好多参考了ai和网络上的答案】，会推荐看这个还是jyy的os?目的是后续找后端开发这样的方向的

**Reply by Zlatanwic** [reply]
普通后端开发没必要学这么深

**Reply by SweetGargamel** [reply]
意思是说这俩课程都不用听吗，直接去看八股吗

**Reply by Tinuvile** [reply]
@SweetGargamel 找后端开发的话建议早点做业务类的项目，我也是大二在这些课上花了很多时间，但这些还是对找infra的工作有用一些

**Comment by rinevard** [comment]
我完成了 2024 版本的 6S081，把 Lab 笔记和代码存放在了 [这个 github 仓库](https://github.com/rinevard/MIT-6.828-2024) 中。我在学习过程中用到的最重要的资料是 [6S081-translation](https://mit-public-courses-cn-translatio.gitbook.io/)，这是每节课的中文文字版，看这个和看视频效果差不多。这是2020版本的视频文字版，和当前的代码有少许不一致，但基本不影响学习。另外一些资料我放到了 [蓝奏云](https://rinevard.lanzn.com/ioHMA3d1p93a) 中（也可以在上述的 github 库中下载），包括但不限于教材双语对照机翻、RISC-V 手册的中文版（出处为 [这里]( https://github.com/ISRC-CAS/riscv-isa-manual-cn)）

**Comment by weilaistar** [comment]
有没有一起做的，在这蹲一个（最新版从头开始，QQ：1274008479）

**Comment by in-people** [comment]
打算26年前三个月完成OS实验，欢迎小伙伴一起。

**Reply by in-people** [reply]
QQ 709065175

**Comment by glasscatya** [comment]
课程视频翻译文档挂了，我补档下。 https://zh.cs.mit.glasscat.top/library/mit-s081/

**Comment by KDBRonaldo** [comment]
请问这门课2021fall之后没有公开的课程网站了吗？

**Reply by 114514kirito** [reply]
有，最新的是25年的，改名为mit6.1810了，https://pdos.csail.mit.edu/6.1810/2025/

**Reply by Arachniddd** [reply]
这个课程网站最近是不是关闭了？之前还打得开最近打不开了
